contents/blog/6-learnings-from-1-year-of-agents-posthog-ai.mdx

```diff
@@ -0,0 +1,95 @@
+
+## 4. Wider context is key
+
+Did I mention context matters? It really does, at every step of a task, because people define tasks in the most ambiguous ways. How would _you_ answer "where do users drop off in cfmp conversion"? It looks like a bunch of typos sneaked in - unless you know CFMP is a product offerred by the user, and then you need to understand how it fits into their world (this is a real example). What your agent needs is a `CLAUDE.md` equivalent – core context that's always attached.
+
+Core context massively improves agent trajectories, but it's a new to-do for a user, and none us love it when a new tool creates setup work. We've found it critical for core context creation to be effortless, yet not forced. Ultimately, as all great artists do, we stole from the best – and implemented the `/init` command, snatched from Claude Code.
+
+In PostHog AI, `/init` learns about your product and business via multi-step web search (currently using GPT 5 mini). The results form the agent's project-level memory. In the case your data doesn't contain URLs to work with, a few questions are asked directly, but it's something we try to minimize.
+
+Context is _everywhere_ though. It lives on the web, but it lives in your Slack too, in the email inbox, in your notes app. This is one of the problems in applied AI that already have a number of solutions, but aren't _solved_ yet.
+
+## 5. Frameworks considered harmful
+
+Early on, we migrated from the OpenAI Python SDK to LangChain + LangGraph. Today, we absolutely wouldn't.
+
+Remember the memes that JavaScript gets a new frontend framework every second? That's LLMs now, only LLMs evolve much faster than JS ever did, and the frameworks deliver less value.
+
+The elegance of LiteLLM or Vercel AI crumbles when AI providers add new models or new features. Just check out the mess that web search results are, as OpenAI and Anthropic format them completely differently, while the frameworks try to maintain one facade.
+
+But never, _ever_ store your state with an opinionated framework. Roll your own state. The orchestration layer of a framework like LangGraph – fine. Perhaps it's [obsolete now](#agents-beat-workflows), but it's just a way of calling functions. The state management layer of LangGraph – little-death that brings total obliteration. A custom format that relies on Python's pickling format is the mind-killer. These sorts of frameworks are fun at first, then a pain to migrate from.
+
+AI may settle upon its React one day, but for now the framework wars rage on. It might just be best to stay low-level.
+
+## 6. Evals are not nearly all you need
+
+[Some](https://x.com/gdb/status/1733553161884127435) say evals can be all you need. For foundations models: certainly, you should curate your datasets. For agents: you should curate datasets too! We've found evals important for making meaningful changes with base confidence.
```

Comment:
This reads like this:
- Some say evals are all you need (hinting that it's not this way)
- This is certainly true for foundational models (thesis)
- and it's true for agents too (missed antithesis, here I would have expected you to say: "but for agents, it's a bit more nuanced", and then move to the next paragraph where you explain the synthesis)